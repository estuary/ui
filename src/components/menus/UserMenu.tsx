/* eslint-disable @typescript-eslint/no-unnecessary-condition */
import type { SxProps } from '@mui/material';

import { Stack, Typography } from '@mui/material';
import Divider from '@mui/material/Divider';
import ListItemIcon from '@mui/material/ListItemIcon';
import MenuItem from '@mui/material/MenuItem';

import { useShallow } from 'zustand/react/shallow';

import { LogOut, Mail, ProfileCircle } from 'iconoir-react';
import { FormattedMessage, useIntl } from 'react-intl';

import useEnhancedSupport from 'src/_compliance/hooks/useEnhancedSupport';
import IconMenu from 'src/components/menus/IconMenu';
import UserAvatar from 'src/components/shared/UserAvatar';
import { supabaseClient } from 'src/context/GlobalProviders';
import { useUserStore } from 'src/context/User/useUserContextStore';

interface Props {
    iconColor: string;
}

const nonInteractiveMenuStyling: SxProps = {
    '&:hover': {
        cursor: 'revert',
    },
};

const UserMenu = ({ iconColor }: Props) => {
    const intl = useIntl();
    const userDetails = useUserStore(useShallow((state) => state.userDetails));

    const { setEnhancedSupport: toggleEnhancedSupport } = useEnhancedSupport();

    const handlers = {
        logout: async () => {
            await supabaseClient.auth.signOut();
        },
    };

    if (userDetails) {
        const { avatar, email, emailVerified, userName } = userDetails;
        return (
            <IconMenu
                ariaLabel={intl.formatMessage({ id: 'accountMenu.ariaLabel' })}
                icon={
                    <UserAvatar
                        userEmail={email}
                        userName={userName}
                        avatarUrl={avatar}
                    />
                }
                identifier="account-menu"
                tooltip={intl.formatMessage({ id: 'accountMenu.tooltip' })}
            >
                <MenuItem sx={nonInteractiveMenuStyling}>
                    <ListItemIcon>
                        <ProfileCircle style={{ color: iconColor }} />
                    </ListItemIcon>
                    {userName}
                </MenuItem>

                <MenuItem sx={nonInteractiveMenuStyling}>
                    <ListItemIcon>
                        <Mail style={{ color: iconColor }} />
                    </ListItemIcon>

                    <Stack spacing={0}>
                        <Typography>{email}</Typography>

                        {emailVerified ? (
                            <Typography variant="caption">
                                <FormattedMessage id="accountMenu.emailVerified" />
                            </Typography>
                        ) : null}
                    </Stack>
                </MenuItem>

                <Divider />

                <MenuItem
                    onClick={() => {
                        void handlers.logout();
                    }}
                >
                    <ListItemIcon>
                        <LogOut style={{ color: iconColor }} />
                    </ListItemIcon>

                    <FormattedMessage id="cta.logout" />
                </MenuItem>

                <MenuItem
                    onClick={() => {
                        toggleEnhancedSupport(true);
                    }}
                >
                    Enable LogRocket
                </MenuItem>
            </IconMenu>
        );
    } else {
        return null;
    }
};

export default UserMenu;
