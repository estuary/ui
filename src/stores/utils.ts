import { isEmpty, map } from 'lodash';

export const checkForErrors = (obj: any) =>
    obj?.errors && obj.errors.length > 0;

export const populateErrors = (configDictionary: any) => {
    let configErrors: any[] = [];

    // TODO (errors) When there are not configs do we need to populate this object with something?
    if (Object.keys(configDictionary).length > 0) {
        map(configDictionary, (config) => {
            if (checkForErrors(config)) {
                configErrors = configErrors.concat(config.errors);
            }
        });
    }

    return {
        configErrors,
        hasErrors: !isEmpty(configErrors),
    };
};
