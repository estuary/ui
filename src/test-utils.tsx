// https://testing-library.com/docs/react-testing-library/setup#custom-render

import { render, RenderOptions } from '@testing-library/react';
import { ReactElement, type FC } from 'react';
import { act } from 'react-dom/test-utils';
import { localStorageKey } from './auth';
import AppProviders from './context';

const AllTheProviders: FC = ({ children }) => {
    return <AppProviders>{children}</AppProviders>;
};

const customRender = (
    ui: ReactElement,
    options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

const loginAsUser = (userName: string = 'fakeUserName') => {
    act(() => {
        window.localStorage.setItem(localStorageKey, userName);
    });
};

export * from '@testing-library/react';
export { customRender as render, loginAsUser };
