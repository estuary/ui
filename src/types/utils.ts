export type WithRequiredProperty<T, K extends keyof T> = T & {
    [P in K]-?: T[P];
};

export type WithRequiredNonNullProperty<T, K extends keyof T> = T & {
    [P in K]-?: Exclude<T[P], null>;
};
