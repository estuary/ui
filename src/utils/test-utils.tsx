/* eslint-disable @typescript-eslint/unbound-method */
// https://testing-library.com/docs/react-testing-library/setup#custom-render

import { render as rtlRender, RenderOptions } from '@testing-library/react';
import { AuthContext } from 'context/Auth';
import AppContent from 'context/Content';
import AppRouter from 'context/Router';
import AppTheme from 'context/Theme';
import { ReactElement } from 'react';
import { auth, AuthDetails } from 'services/auth';

// TODO - this does not really do anything because we hardcode
//   the username down below on the MockAuthProvider. However,
//   this will be needed in the future so leaving the work so far
//   so we can leverage this when wanting to test the "on load auth"
//   functionality
const loginAsUser = (username: string) => {
    const mockAuthDetails: AuthDetails = {
        session: {
            account_id: '',
            expires_at: '',
            token: '',
        },
        user: {
            created_at: '',
            display_name: username,
            email: `${username}@${username}`,
            id: '',
            name: username,
            unique_name: username,
            updated_at: '',
        },
    };

    auth.saveSession(mockAuthDetails.session);
    auth.saveUser(mockAuthDetails.user);
};

const logoutUser = () => {
    auth.removeAuthDetails();
};

const goTo = (route?: string, name?: string) => {
    window.history.pushState(
        {},
        name ? name : 'Test page',
        route ? route : '/'
    );
};

const waitForLoading = async () => {
    return Promise.resolve();
};

// Mocking AuthContext out basically skips all the loading screen and boostrapping
//  stuff that happens on load. This should be fine for unit testing as it
//  can help speed up tests a bit and makes things easier to deal with. This
//  does mean that we need to eventually write a basic test that handles the
//  account bootstrapping.
interface MockAuthProps {
    children: JSX.Element;
    username?: string;
}
const MockAuthProvider = ({ children, username }: MockAuthProps) => {
    const value = {
        login: jest.fn(() => {
            if (username) {
                loginAsUser(username);
            }
            return Promise.resolve();
        }),
        logout: jest.fn(() => {
            logoutUser();
            return Promise.resolve();
        }),
        user: username,
    };

    return (
        <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
    );
};

const customRender = async (
    ui: ReactElement,
    options: Omit<RenderOptions, 'wrapper'> & {
        route?: string;
        user?: string;
    }
) => {
    const { route, user } = options;

    goTo(route, 'Test Page');

    const view = rtlRender(
        <AppContent>
            <MockAuthProvider username={user}>
                <AppTheme>
                    <AppRouter>{ui}</AppRouter>
                </AppTheme>
            </MockAuthProvider>
        </AppContent>,
        {
            ...options,
        }
    );

    await waitForLoading();

    return view;
};

export * from '@testing-library/react';
export { goTo, loginAsUser, logoutUser, customRender, waitForLoading };
