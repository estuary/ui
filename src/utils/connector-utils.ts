// TODO (Typing)
// Since the typing looks at columns it was a pain to make this
//  truly reusable. So marking the query as `any` even thogh
//  it is PostgrestFilterBuilder<ConnectorTag |ConnectorWithTagDetailQuery>

import type { PostgrestFilterBuilder } from '@supabase/postgrest-js';
import type { ConnectorConfig } from 'deps/flow/flow';
import type {
    ConnectorsQuery_DetailsForm,
    ConnectorTag_Base,
} from 'src/api/connectors';
import type { DraftSpecsExtQuery_ByDraftId } from 'src/api/draftSpecs';
import type { ConnectorWithTagDetailQuery } from 'src/hooks/connectors/shared';
import type { LiveSpecsExtQuery } from 'src/hooks/useLiveSpecsExt';
import type {
    ConnectorMetadata,
    DekafConnectorMetadata,
    Details,
    StandardConnectorMetadata,
} from 'src/stores/DetailsForm/types';
import type { DekafConfig } from 'src/types';

import { hasLength } from 'src/utils/misc-utils';

const DEKAF_IMAGE_PREFIX = 'ghcr.io/estuary/dekaf-';
const DEKAF_VARIANT_PROPERTY = 'variant';

export const isDekafConnector = (
    value: StandardConnectorMetadata | DekafConnectorMetadata
): value is DekafConnectorMetadata => DEKAF_VARIANT_PROPERTY in value;

export const isDekafEndpointConfig = (
    value: ConnectorConfig | DekafConfig
): value is DekafConfig => DEKAF_VARIANT_PROPERTY in value;

// TODO (typing): Align `connectors` and `connector_tags` query interfaces.
//   Renamed table columns need to be given the same name to avoid type conflicts.
export function getConnectorMetadata(
    connector: ConnectorsQuery_DetailsForm | ConnectorWithTagDetailQuery,
    options?: ConnectorVersionEvaluationOptions
): Details['data']['connectorImage'] {
    const { id: connectorTagId, image_tag } = evaluateConnectorVersions(
        connector,
        options
    );

    const { id: connectorId, image: iconPath, image_name } = connector;

    const connectorMetadata: ConnectorMetadata = {
        connectorId,
        iconPath,
        id: connectorTagId,
        imageName: image_name,
        imageTag: image_tag,
    };

    return image_name.startsWith(DEKAF_IMAGE_PREFIX)
        ? {
              ...connectorMetadata,
              variant: image_name.substring(DEKAF_IMAGE_PREFIX.length),
          }
        : {
              ...connectorMetadata,
              imagePath: `${image_name}${image_tag}`,
          };
}

export const getEndpointConfig = (
    data: DraftSpecsExtQuery_ByDraftId[] | LiveSpecsExtQuery[]
) =>
    Boolean(data[0].spec.endpoint?.dekaf)
        ? data[0].spec.endpoint.dekaf.config
        : data[0].spec.endpoint.connector.config;

// TODO (V2 typing) - query should take in filter builder better
export const requiredConnectorColumnsExist = <Response>(
    query: any,
    columnPrefix?: string
): PostgrestFilterBuilder<any, any, Response> => {
    return query
        .not(`${columnPrefix ? `${columnPrefix}.` : ''}image_tag`, 'is', null)
        .not(
            `${columnPrefix ? `${columnPrefix}.` : ''}resource_spec_schema`,
            'is',
            null
        )
        .not(
            `${columnPrefix ? `${columnPrefix}.` : ''}endpoint_spec_schema`,
            'is',
            null
        );
};

export interface ConnectorVersionEvaluationOptions {
    connectorId: string;
    existingImageTag: string;
}

export function evaluateConnectorVersions(
    connector: ConnectorWithTagDetailQuery | ConnectorsQuery_DetailsForm,
    options?: ConnectorVersionEvaluationOptions
): ConnectorTag_Base {
    // Return the version of the connector that is used by the existing task in an edit workflow.
    if (options && options.connectorId === connector.id) {
        const connectorsInUse = connector.connector_tags.filter(
            (version) => version.image_tag === options.existingImageTag
        );

        if (hasLength(connectorsInUse)) {
            return connectorsInUse[0];
        }
    }

    // Return the latest version of a given connector.
    const { connector_id, id, image_tag } = connector.connector_tags.sort(
        (a, b) => b.image_tag.localeCompare(a.image_tag)
    )[0];

    return { connector_id, id, image_tag };
}
