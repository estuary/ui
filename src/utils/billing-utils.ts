import { isEqual, parseISO } from 'date-fns';
import { isEmpty, sum } from 'lodash';
import prettyBytes from 'pretty-bytes';
import { BillingRecord } from 'stores/Billing/types';
import { CatalogStats_Billing, Entity } from 'types';

export const TOTAL_CARD_HEIGHT = 300;

// Grid item height - 72 = graph canvas height
export const CARD_AREA_HEIGHT = TOTAL_CARD_HEIGHT - 72;

export const BYTES_PER_GB = 1073741824;

const FREE_TASK_COUNT = 2;

export enum FREE_GB_BY_TIER {
    FREE = 20,
    PERSONAL = 10,
    ENTERPRISE = 10,
}

export const evaluateSpecType = (query: CatalogStats_Billing): Entity => {
    if (Object.hasOwn(query.flow_document.taskStats, 'capture')) {
        return 'capture';
    } else if (Object.hasOwn(query.flow_document.taskStats, 'materialize')) {
        return 'materialization';
    } else {
        return 'collection';
    }
};

const evaluateTotalCost = (dataVolume: number, taskCount: number) => {
    const freeBytes = 10 * BYTES_PER_GB;

    const dataVolumeOverLimit =
        dataVolume > freeBytes ? dataVolume - freeBytes : 0;

    const taskCountOverLimit =
        taskCount > FREE_TASK_COUNT ? taskCount - FREE_TASK_COUNT : 0;

    return (
        (dataVolumeOverLimit / BYTES_PER_GB) * 0.75 + taskCountOverLimit * 20
    );
};

const evaluateDataVolume = (stats: CatalogStats_Billing[]): number => {
    const taskBytes: number[] = stats.map((query) => {
        if (Object.hasOwn(query.flow_document.taskStats, 'capture')) {
            return query.bytes_written_by_me;
        } else if (
            Object.hasOwn(query.flow_document.taskStats, 'materialize')
        ) {
            return query.bytes_read_by_me;
        } else {
            return query.bytes_written_by_me + query.bytes_read_by_me;
        }
    });

    return sum(taskBytes);
};

export const stripTimeFromDate = (date: string) => {
    const [truncatedDateStr] = date.split('T', 1);

    return parseISO(truncatedDateStr);
};

const getInitialBillingRecord = (date: string): BillingRecord => {
    const truncatedDate = stripTimeFromDate(date);

    return {
        date: truncatedDate,
        dataVolume: 0,
        taskCount: 0,
        totalCost: 0,
        pricingTier: null,
        taskRate: null,
        gbFree: null,
        includedTasks: null,
    };
};

// TODO (billing): Remove this helper function to translate data returned from
//   the new RPC when available.
export const formatBillingCatalogStats = (
    value: CatalogStats_Billing[]
): BillingRecord[] => {
    const taskStatData = value.filter((query) =>
        Object.hasOwn(query.flow_document, 'taskStats')
    );

    let sortedStats: {
        [date: string]: CatalogStats_Billing[];
    } = {};

    taskStatData.forEach((query) => {
        if (Object.hasOwn(sortedStats, query.ts)) {
            sortedStats[query.ts].push(query);
        } else {
            sortedStats = {
                ...sortedStats,
                [query.ts]: [query],
            };
        }
    });

    const billingHistory: BillingRecord[] = [];

    if (!isEmpty(sortedStats)) {
        Object.entries(sortedStats).forEach(([ts, stats]) => {
            const billingRecordIndex = billingHistory.findIndex((record) =>
                isEqual(record.date, stripTimeFromDate(ts))
            );

            const taskCount = stats.length;
            const dataVolume = evaluateDataVolume(stats);
            const totalCost = evaluateTotalCost(dataVolume, taskCount);

            if (billingRecordIndex === -1) {
                const { date, pricingTier, taskRate, gbFree, includedTasks } =
                    getInitialBillingRecord(ts);

                billingHistory.push({
                    date,
                    dataVolume,
                    taskCount,
                    totalCost,
                    pricingTier: pricingTier ?? 'personal',
                    taskRate: taskRate ?? 20,
                    gbFree: gbFree ?? FREE_GB_BY_TIER.PERSONAL,
                    includedTasks: includedTasks ?? 2,
                });
            } else {
                const { date, pricingTier, taskRate, gbFree, includedTasks } =
                    billingHistory[billingRecordIndex];

                billingHistory[billingRecordIndex] = {
                    date,
                    dataVolume,
                    taskCount,
                    totalCost,
                    pricingTier: pricingTier ?? 'personal',
                    taskRate: taskRate ?? 20,
                    gbFree: gbFree ?? FREE_GB_BY_TIER.PERSONAL,
                    includedTasks: includedTasks ?? 2,
                };
            }
        });
    }

    return billingHistory;
};

export interface SeriesConfig {
    data: [string, number][];
    seriesName?: string;
    stack?: string;
}

export const formatDataVolumeForDisplay = (
    seriesConfigs: SeriesConfig[],
    tooltipConfig: any
): string => {
    const filteredSeries =
        seriesConfigs.length === 1
            ? seriesConfigs
            : seriesConfigs.filter(
                  (series) => series.seriesName === tooltipConfig.seriesName
              );

    const dataVolumeInBytes = filteredSeries
        .flatMap(({ data }) => data)
        .find(([month]) => month === tooltipConfig.name);

    return dataVolumeInBytes
        ? prettyBytes(dataVolumeInBytes[1], { minimumFractionDigits: 2 })
        : `${tooltipConfig.value[1]} GB`;
};
