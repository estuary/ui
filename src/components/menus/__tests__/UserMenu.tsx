import '@testing-library/jest-dom';
import { UserEvent } from '@testing-library/user-event/dist/types/setup/setup';
import UserMenu from 'components/menus/UserMenu';
import { act, customRender, screen, waitFor } from 'utils/test-utils';

describe('When there is user metadata', () => {
    const openMenu = async (user: UserEvent) => {
        await user.click(
            screen.getByLabelText<HTMLImageElement>('Open Account Menu')
        );
    };

    let username: string;

    beforeEach(() => {
        username = 'foo123';
    });

    describe('the menu displays', () => {
        test('full name', async () => {
            const { user } = await customRender(
                <UserMenu iconColor="black" />,
                {
                    username,
                }
            );

            await act(async () => {
                await openMenu(user);
            });

            await waitFor(async () => {
                expect(
                    screen.getByText(`Full ${username}`)
                ).toBeInTheDocument();
            });
        });

        test('email', async () => {
            const { user } = await customRender(
                <UserMenu iconColor="black" />,
                {
                    username,
                }
            );

            await act(async () => {
                await openMenu(user);
            });

            await waitFor(async () => {
                expect(
                    screen.getByText(`${username}@example.org`)
                ).toBeInTheDocument();
            });
        });
    });
});
