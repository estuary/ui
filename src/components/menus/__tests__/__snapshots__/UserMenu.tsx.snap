// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When there is user metadata the menu displays snapshot 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <button
      aria-controls="account-menu-menu"
      aria-expanded="true"
      aria-haspopup="true"
      aria-label="Open account menu"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-j8rmhe-MuiButtonBase-root-MuiIconButton-root"
      data-mui-internal-clone-element="true"
      id="account-menu-button"
      tabindex="0"
      type="button"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular css-10zwf3c-MuiAvatar-root"
      >
        <img
          class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
          src="https://example.org/avatar/foo123"
        />
      </div>
    </button>
  </div>
  <div
    aria-labelledby="account-menu-button"
    class="MuiModal-root MuiPopover-root MuiMenu-root css-10nakn3-MuiModal-root-MuiPopover-root-MuiMenu-root"
    id="account-menu-menu"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiBackdrop-invisible css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiMenu-paper MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiPopover-paper css-mz8czf-MuiPaper-root-MuiMenu-paper-MuiPaper-root-MuiPopover-paper"
      style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
      tabindex="-1"
    >
      <ul
        class="MuiList-root MuiList-padding MuiMenu-list css-6hp17o-MuiList-root-MuiMenu-list"
        role="menu"
        tabindex="-1"
      >
        <li
          class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-1qldtwa-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="AccountCircleIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
            </svg>
          </div>
          Full foo123
        </li>
        <li
          class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-1qldtwa-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="-1"
        >
          <div
            class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="EmailIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4-8 5-8-5V6l8 5 8-5v2z"
              />
            </svg>
          </div>
          <div
            class="css-nen11g-MuiStack-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1q5p0sd-MuiTypography-root"
            >
              foo123@example.org
            </p>
            <span
              class="MuiTypography-root MuiTypography-caption css-eiw59m-MuiTypography-root"
            >
              verified
            </span>
          </div>
        </li>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
        />
        <li
          class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-1qldtwa-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="-1"
        >
          <div
            class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="LogoutIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m17 7-1.41 1.41L18.17 11H8v2h10.17l-2.58 2.58L17 17l5-5zM4 5h8V3H4c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h8v-2H4V5z"
              />
            </svg>
          </div>
          Logout
        </li>
      </ul>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
